### ISSUE TITLE
Documents and Topics are different lengths and cannot merge the topics

### ISSUE THREAD

**Issue by daianacric95** [MaartenGr/BERTopic#1626]
Hi there Maarten!

Thank you for all of the support you are offering the community!

I am running into the problem that my docs and topics have different lengths after I ran the model. I do not have the possibility to run the model again, but because of this issue, I cannot plot the topics over time or merge them. 

Would it be possible to suggest how to proceed from here?

<img width="1218" alt="Screenshot 2023-11-10 at 14 35 04" src="https://github.com/MaartenGr/BERTopic/assets/28362483/6bcc9c34-4cb7-4e68-98fd-67281de2b126">

**Comment by MaartenGr** [comment]
Did you make sure that the number of documents in `text_col` is the same as the number of dates in `date`? The only way to run `.topics_over_time` is by making sure that both the documents and the dates are the same size. I am assuming you have the raw data saved, so why not use that as the input for `topics_over_time`?

**Comment by daianacric95** [comment]
Hi there, Maarten! Thank you so much for taking the time to answer my question. I have realized that I made a mistake during the batch processing, and I have not updated the topics as you suggested in the Online Topic Modeling section. Once I fixed that, the length of the topics and the documents is the same and I could plot the hierarchical topics. Although the problem that I am encountering now, is when I try to merge topics with the following code:         
`topics_to_merge = [1, 2]
topic_model.merge_topics(docs, topics_to_merge)` 
I am getting a KeyError and I am not able to merge the topics. 
I am attaching the batch-processing code for further context:
`
indices = np.arange(len(text_column))
np.random.shuffle(indices)
text_column = [text_column[i] for i in indices]

# Divide text_column and loaded_embeddings into chunks
chunk_size = 10000
text_chunks = [text_column[i:i + chunk_size] for i in range(0, len(text_column), chunk_size)]

# Process each chunk separately
topics = []

for i in tqdm(range(len(text_chunks)), desc="Processing chunks", ):
    text_chunk = text_chunks[i]
    topics_chunk = model.fit_transform(text_chunk)
    topics.extend(topics_chunk)

topic_model.topics_ = topics   
  `

**Comment by MaartenGr** [comment]
Could you format your code with ```python so that it is easier to read? Also, could you share your full code? Moreover, could you share the **full** error that you received? Without it, it is difficult to understand where the issue lies.

**Comment by daianacric95** [comment]
Hello Maarten! Thank you for being so patient. Here is the full I have used and the error I am getting when trying to merge the topics:

```
umap_model = UMAP()
hdbscan_model = HDBSCAN()
vectorizer_model = CountVectorizer(stop_words="english")
embedding_model = SentenceTransformer('all-MiniLM-L6-v2')

topic_model = BERTopic(
    umap_model=umap_model,
    hdbscan_model=hdbscan_model,
    embedding_model=embedding_model,
    vectorizer_model=vectorizer_model,
    top_n_words=10,
    language='english',
    verbose=True
)

# Shuffle data
indices = np.arange(len(text_col))
np.random.shuffle(indices)
text_col = [text_col[i] for i in indices]

# Divide text_column and loaded_embeddings into chunks
chunk_size = 1000
text_chunks = [text_col[i:i + chunk_size] for i in range(0, len(text_col), chunk_size)]

# Process each chunk separately
topics = []
for i in range(len(text_chunks)):
    text_chunk = text_chunks[i]
    topics_chunk = topic_model.fit_transform(text_chunk)
    topics.extend(topic_model.topics_)

topic_model.topics_ = topics

## Merge topics

topics_to_merge = [2, 3]
topic_model.merge_topics(text_col, topics_to_merge)
```
<img width="1215" alt="Screenshot 2023-11-18 at 11 55 54" src="https://github.com/MaartenGr/BERTopic/assets/28362483/cdd90181-49c9-423c-8b6f-e1a0b782deb6">

**Comment by daianacric95** [comment]
Update: I managed to make River work (needed to install an older version) but I am once again encountering the same issue:

```
from river import stream
from river import cluster

class River:
    def __init__(self, model):
        self.model = model

    def partial_fit(self, umap_embeddings):
        for umap_embedding, _ in stream.iter_array(umap_embeddings):
            self.model = self.model.learn_one(umap_embedding)

        labels = []
        for umap_embedding, _ in stream.iter_array(umap_embeddings):
            label = self.model.predict_one(umap_embedding)
            labels.append(label)

        self.labels_ = labels
        return self

# Using DBSTREAM to detect new topics as they come in
cluster_model = River(cluster.DBSTREAM())
vectorizer_model = OnlineCountVectorizer(stop_words="english")
ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True, bm25_weighting=True)

# Prepare model
topic_model_v2 = BERTopic(
    hdbscan_model=cluster_model, 
    vectorizer_model=vectorizer_model, 
    ctfidf_model=ctfidf_model,
)

batch_size = 1000
batches = [text_col[i:i + batch_size] for i in range(0, len(text_col), batch_size)]
all_my_topics=[]
# Incrementally fit the topic model by training on 1000 documents at a time
for batches in batches:
    topic_model_v2.partial_fit(batch)
```
<img width="1222" alt="Screenshot 2023-11-18 at 22 25 54" src="https://github.com/MaartenGr/BERTopic/assets/28362483/c049b1cf-3e18-441e-a084-97f242e3b39c">

Additionally the results of the partial_fit are substantially different than the ones of fit_transform(), where I get 200 topics, while with partial_fit is only 18

**Comment by MaartenGr** [comment]
In your code:

```python
batch_size = 1000
batches = [text_col[i:i + batch_size] for i in range(0, len(text_col), batch_size)]
all_my_topics=[]
# Incrementally fit the topic model by training on 1000 documents at a time
for batches in batches:
    topic_model_v2.partial_fit(batch)
```

You are not updating the internal `topic_model_v2.topics_` which still should be done. So something like this:

```python
batch_size = 1000
batches = [text_col[i:i + batch_size] for i in range(0, len(text_col), batch_size)]
all_my_topics=[]
# Incrementally fit the topic model by training on 1000 documents at a time
for batches in batches:
    topic_model_v2.partial_fit(batch)
    all_my_topics.extend(topic_model_v2.topics_)

topic_model_v2.topics_ = all_my_topics
```

> Additionally the results of the partial_fit are substantially different than the ones of fit_transform(), where I get 200 topics, while with partial_fit is only 18

This is to be expected since you are using two completely different clustering models. In the former,m you are using HDBSCAN whereas the latter uses DBSTREAM. Moreover, the training procedures are different since in the former you are training UMAP on the entire dataset whilst in the latter you are training UMAP only on the very first batch.

Instead, your use case might be good for the newly introduced `.merge_models`. The method allows for different topic models to be merged together. When you combine two models with this method, the first model will remain as it is and the second model will be added as long as it contains new clusters. Existing clusters will not be added since those were already found in the first model.

You can do this continuously and keep on merging models this way every time you train a new model. This means that you can use this method for incremental learning by iteratively training and emerging models. You can find more about that https://github.com/MaartenGr/BERTopic/pull/1516.

**Comment by MaartenGr** [comment]
Could you share your full code with the suggestion you tried? That makes communication a bit easier.
